docs/developers/testing.md

```diff
@@ -0,0 +1,314 @@
+```
+virtualenv -p python3 ~/irods_testing_environment
+source ~/irods_testing_environment/bin/activate
+pip install docker-compose GitPython
+pip freeze
+```
+
+#### If you cannot install `docker-compose` ...
+
+The iRODS Testing Environment is built on top of the now-defunct Python implementation of Docker Compose called `docker-compose`: [https://github.com/docker/compose/tree/1.29.2](https://github.com/docker/compose/tree/1.29.2) `docker-compose` is being phased out by Docker and you may experience problems installing it via `pip`.
+
+If this happens to you, try doing the following:
+
+Clone the Docker Compose Git repository:
+```bash
+git clone https://github.com/docker/compose
+```
+
+Check out the latest tag which was still using the Python implementation:
+```bash
+cd compose
+git checkout 1.29.2
+```
+
+At this point, you can make the modifications needed to fix any problems you may encounter.
+
+Once done, back out and pip install the local directory:
+```bash
+cd -
+pip install ./compose
+```
+
+### How to run core tests: `run_core_tests.py`
+
+`run_core_tests.py` is the main script for running the iRODS "core" test suite (see [Core test suite](#core-test-suite)).
+
+Here is the minimal form for running the test suite for a particular `OS`, `OS_VERSION`, `DATABASE`, and `DATABASE_VERSION`:
+```
+python run_core_tests.py --project-directory projects/OS-OS_VERSION/OS-OS_VERSION-DATABASE-DATABASE_VERSION
+```
+
+This will serially run the entire test suite using the latest release of the iRODS server. This will create a container for the iRODS server configured as the sole catalog service provider, a container for the database where the catalog is stored, and a network so that they can communicate as if it were a distributed environment.
+
+#### Using locally built iRODS packages
+
+The `--irods-package-directory` option allows the user to provide iRODS packages as files on the host machine. This is the standard way for iRODS developers to test modifications to the iRODS server software:
+```
+python run_core_tests.py \
+    --project-directory projects/OS-OS_VERSION/OS-OS_VERSION-DATABASE-DATABASE_VERSION \
+    --irods-package-directory /path/to/directory/containing/irods/package/files
+```
+
+#### Using concurrency to reduce test time
+
+The `--concurrent-test-executors` option allows the user to specify how many iRODS zones to deploy and use for splitting up the list of tests to run. Note that this option can quickly increase the number of containers created by the Compose project (2 containers per test executor), so make sure to choose a number appropriate to the host system's capabilities. Here is an example of how to use the option:
+```
+python run_core_tests.py \
+    --project-directory projects/OS-OS_VERSION/OS-OS_VERSION-DATABASE-DATABASE_VERSION \
```

Comment:
i might be coming around to your style... it's less verbose... and once you see it correctly, you can 'ignore' it...